miri-script/src/commands.rs

```diff
@@ -0,0 +1,711 @@
+}
+
+fn apply_env_variables<'a>(
+    cmd: &'a mut duct::Expression,
+    variables: &[(String, String)],
+) -> &'a mut duct::Expression {
+    for (key, value) in variables {
+        *cmd = cmd.env(key, value);
+    }
+    cmd
+}
+#[derive(Clone, Debug)]
+pub(super) struct MiriRunner<'a> {
+    /// miri_dir is based on build directory of miri-script.
+    miri_dir: PathBuf,
+    /// active_toolchain is passed as `+toolchain` argument to cargo/rustc invocations.
+    active_toolchain: String,
+    cargo_extra_flags: String,
+    command: &'a super::Subcommands,
+    /// Environment variables passed to child processes.
+    extra_flags: BTreeMap<String, String>,
+    /// Additional variables used by environment-altering commands.
+    extra_vars: Extras,
+}
+
+#[derive(Clone, Debug, Default)]
+struct Extras {
+    sysroot: PathBuf,
+    libdir: PathBuf,
+}
+
+impl Extras {
+    fn new(toolchain: &str) -> Result<Self> {
+        // Determine toolchain properties.
+        let toolchain_prop = &format!("+{toolchain}");
+        let target_output = cmd("rustc", [&toolchain_prop, "--version", "--verbose"]).read()?;
+        let host_line = target_output
+            .lines()
+            .find(|line| line.starts_with("host:"))
+            .expect("rustc output must contain host info");
+        let target = host_line.split(' ').nth(1).expect("host line does not contain host info");
+
+        let sysroot = cmd("rustc", [&toolchain_prop, "--print", "sysroot"]).read()?;
+
+        let sysroot = PathBuf::from(sysroot);
+        let libdir = path!(sysroot / "lib" / "rustlib" / target / "lib");
+        Ok(Self { libdir, sysroot })
+    }
+}
+
+impl<'a> MiriRunner<'a> {
+    pub(super) fn run(command: &'a super::Subcommands) -> Result<()> {
+        let miri_dir = detect_miri_dir()?;
+        let active_toolchain = get_current_toolchain(&miri_dir)?;
+        let config = command.get_auto_config(&miri_dir);
+        let cargo_extra_flags = std::env::var("CARGO_EXTRA_FLAGS").unwrap_or_default();
+        let extra_flags = BTreeMap::new();
+
+        let mut ret = Self {
+            miri_dir,
+            active_toolchain,
+            command,
+            extra_flags,
+            extra_vars: Default::default(),
+            cargo_extra_flags,
+        };
+        if let Some(config) = config {
+            ret.extra_flags.insert("MIRI_AUTO_OPS".into(), "42".into());
+            if config.toolchain {
+                ret.toolchain(&[])?;
+            }
+            if config.fmt {
+                ret.fmt(&[])?;
+            }
+            if config.clippy {
+                ret.clippy(&["--".to_owned(), "-D".to_owned(), "warnings".to_owned()])?;
+            }
+            if config.toolchain {
+                // Let's make sure to actually use that toolchain, too.
+                // We do it after running all the auto things though to use a detected toolchain.
+                ret.active_toolchain = "miri".to_owned();
+            }
```

Comment:
This should be inside `if config.toolchain` to match the shell script behavior (in particular it must run before fmt/clippy)

Reply:
Uhh, I'm not sure; correct me if I'm wrong, but [TOOLCHAIN variable is not exported](https://github.com/rust-lang/miri/blob/master/miri#L212), so it shouldn't affect child shells. 

On the other hand, `toolchain` overrides toolchain in current directory, so it might be the case that fmt and clippy use `miri` toolchain indirectly anyways. 

Okay, let's just change this for now.

Reply:
> [TOOLCHAIN variable is not exported](https://github.com/rust-lang/miri/blob/master/miri#L212), so it shouldn't affect child shells.

Ah, fair.

> On the other hand, toolchain overrides toolchain in current directory, so it might be the case that fmt and clippy use miri toolchain indirectly anyways.

Indeed that will be the case.